### ISSUE TITLE
Exception thrown: read access violation, when using the C++ API

### ISSUE THREAD

**Issue by jKnepel** [jpdante/UE4-JoyConDriver#3]
When using the C++ API a read access violation is thrown when trying to use the JoyConModule* after importing the Module like the Wiki page described. Tested in Unreal Versions 4.25.4 and 4.27.2. 
```
void AMyCharacter::BeginPlay()
{
	Super::BeginPlay();

	TArray<IJoyConDriverModule*> JoyConInputApis = IModularFeatures::Get().GetModularFeatureImplementations<IJoyConDriverModule>(IJoyConDriverModule::GetModularFeatureName());
	for (IJoyConDriverModule* JoyConInputApi : JoyConInputApis) {
		if (JoyConInputApi == nullptr) continue;
		JoyConModule = JoyConInputApi;
		break;
	}

	if (JoyConModule != nullptr) {
		TArray<FJoyConInformation>* FoundControllers = JoyConModule->SearchForJoyCons();
```
**Unhandled Exception: EXCEPTION_ACCESS_VIOLATION reading address 0xffffffff** on JoyConModule in the last line.

Can be replicated in the provided sample project in the repository. Simply create a C++ class inheriting from character and copy the code from the CppMotionTest Actor that's already in the project or from the wiki and place it in the scene. Access Violation doesn't occur when using the blueprint api instead of the c++ api for these drivers. 
However when using blueprint and opening the project as ndisplay/switchboard application unreal engine will still crash. Below is the Crash Log for ndisplay (only tested in 4.27).

```
[2022.05.02-19.42.19:047][  0]LogWindows: Windows GetLastError: Der Vorgang wurde erfolgreich beendet. (0)
[2022.05.02-19.42.24:631][  0]LogWindows: Error: === Critical error: ===
[2022.05.02-19.42.24:631][  0]LogWindows: Error: 
[2022.05.02-19.42.24:631][  0]LogWindows: Error: Assertion failed: IsValid() [File:C:\ProgramData\EpicGames\UE_4.27\Engine\Source\Runtime\Core\Public\Templates/SharedPointer.h] [Line: 890] 
[2022.05.02-19.42.24:631][  0]LogWindows: Error: 
[2022.05.02-19.42.24:631][  0]LogWindows: Error: 
[2022.05.02-19.42.24:631][  0]LogWindows: Error: 
[2022.05.02-19.42.24:631][  0]LogWindows: Error: [Callstack] 0x00007ffa4d974f69 KERNELBASE.dll!UnknownFunction []
[2022.05.02-19.42.24:631][  0]LogWindows: Error: [Callstack] 0x00007ff9f1ee0216 UE4Editor-Core.dll!ReportAssert() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Windows\WindowsPlatformCrashContext.cpp:1644]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f1ee4218 UE4Editor-Core.dll!FWindowsErrorOutputDevice::Serialize() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Windows\WindowsErrorOutputDevice.cpp:78]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f1c0242d UE4Editor-Core.dll!FOutputDevice::LogfImpl() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Misc\OutputDevice.cpp:61]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f1b56a45 UE4Editor-Core.dll!AssertFailedImplV() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Misc\AssertionMacros.cpp:104]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f1b580a0 UE4Editor-Core.dll!FDebug::CheckVerifyFailedImpl() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Misc\AssertionMacros.cpp:461]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ae77f6ce UE4Editor-JoyConDriver-0001.dll!FJoyConDriverModule::ResumeJoyConConnection() [C:\Users\...\Documents\Git\UnrealCave\Plugins\JoyConDriver\Source\JoyConDriver\Private\JoyConDriverModule.cpp:28]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ae77f825 UE4Editor-JoyConDriver-0001.dll!UJoyConDriverFunctionLibrary::ResumeJoyConConnection() [C:\Users\...\Documents\Git\UnrealCave\Plugins\JoyConDriver\Source\JoyConDriver\Private\JoyConDriverFunctionLibrary.cpp:14]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ae783386 UE4Editor-JoyConDriver-0001.dll!UJoyConDriverFunctionLibrary::execResumeJoyConConnection() [C:\Users\...\Documents\Git\UnrealCave\Plugins\JoyConDriver\Intermediate\Build\Win64\UE4Editor\Inc\JoyConDriver\JoyConDriverFunctionLibrary.gen.cpp:169]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cfbe99 UE4Editor-CoreUObject.dll!UObject::execCallMathFunction() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:956]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cd4d8d UE4Editor-CoreUObject.dll!ProcessLocalScriptFunction() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:1093]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3ca341a UE4Editor-CoreUObject.dll!ProcessScriptFunction<void (__cdecl*)(UObject *,FFrame &,void *)>() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:924]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cd497c UE4Editor-CoreUObject.dll!ProcessLocalFunction() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:1154]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cd4d8d UE4Editor-CoreUObject.dll!ProcessLocalScriptFunction() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:1093]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cd4184 UE4Editor-CoreUObject.dll!UObject::ProcessInternal() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:1181]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f39f19c4 UE4Editor-CoreUObject.dll!UFunction::Invoke() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\Class.cpp:5679]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9f3cd3ca3 UE4Editor-CoreUObject.dll!UObject::ProcessEvent() [D:\Build\++UE4\Sync\Engine\Source\Runtime\CoreUObject\Private\UObject\ScriptCore.cpp:2018]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ee9a942a UE4Editor-Engine.dll!AActor::ProcessEvent() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:863]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ee977435 UE4Editor-Engine.dll!AActor::BeginPlay() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:3579]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ee980c1f UE4Editor-Engine.dll!AActor::DispatchBeginPlay() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:3519]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9efd8ba95 UE4Editor-Engine.dll!AWorldSettings::NotifyBeginPlay() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\WorldSettings.cpp:247]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ef0cca31 UE4Editor-Engine.dll!AGameStateBase::HandleBeginPlay() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\GameStateBase.cpp:205]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9efd2fb3e UE4Editor-Engine.dll!UWorld::BeginPlay() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\World.cpp:4416]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9efc5e611 UE4Editor-Engine.dll!UEngine::LoadMap() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\UnrealEngine.cpp:13264]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9d9d5479c UE4Editor-DisplayCluster.dll!UDisplayClusterGameEngine::LoadMap() [D:\Build\++UE4\Sync\Engine\Plugins\Runtime\nDisplay\Source\DisplayCluster\Private\Game\EngineClasses\Basics\DisplayClusterGameEngine.cpp:275]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9efc089a3 UE4Editor-Engine.dll!UEngine::Browse() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\UnrealEngine.cpp:12463]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff9ef0edbfe UE4Editor-Engine.dll!UGameInstance::StartGameInstance() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\GameInstance.cpp:584]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ea3290 UE4Editor.exe!FEngineLoop::Init() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\LaunchEngineLoop.cpp:4054]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ec0e42 UE4Editor.exe!GuardedMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Launch.cpp:160]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ec10ba UE4Editor.exe!GuardedMainWrapper() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:137]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ec40dd UE4Editor.exe!LaunchWindowsStartup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:273]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ed5984 UE4Editor.exe!WinMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:320]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ff681ed853a UE4Editor.exe!__scrt_common_main_seh() [d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288]
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ffa4e547034 KERNEL32.DLL!UnknownFunction []
[2022.05.02-19.42.24:632][  0]LogWindows: Error: [Callstack] 0x00007ffa4fe62651 ntdll.dll!UnknownFunction []
[2022.05.02-19.42.24:632][  0]LogWindows: Error: 
[2022.05.02-19.42.24:710][  0]LogExit: Executing StaticShutdownAfterError
[2022.05.02-19.42.24:738][  0]LogWindows: FPlatformMisc::RequestExit(1)
[2022.05.02-19.42.24:738][  0]LogWindows: FPlatformMisc::RequestExitWithStatus(1, 3)
[2022.05.02-19.42.24:738][  0]LogCore: Engine exit requested (reason: Win RequestExit)
[2022.05.02-19.42.24:745][  0]Log file closed, 05/02/22 21:42:24
```

**Comment by jpdante** [comment]
Hello, could you provide more information?
What type of project are you using inside Unreal Engine?
Did you tried to play with the provided project? (Not using any extra modes, just how it is)

PS: I never tried to use ndisplay/switchboard and by brief look at the log files it seems it doesn't support custom input drivers.

**Comment by jKnepel** [comment]
Hi, thanks for the fast answer.
After some more experimenting i got the c++ drivers to work in the 4.25.4 project, I'm not sure what caused the previous crashes in this versions, as there was no real difference to what I did now. The provided project works perfectly fine in the older versions now, great job there with the drivers.
The crash still happens in the 4.27.2 project, however this will likely be due to unreal engine version discrepancies. In case you want to debug this (not absolutely required for me, as i can use blueprint instead which works fine in this version), here's how to replicate the crash:

1. clone the repository
2. open the sample project in 4.27.2
3. create a c++ file inheriting from character
4. add the code from the wiki subsection "Importing IJoyConDriverModule" to the c++ file and build.cs file or alternatively simply copy it from the already existing c++ file "CppMotionTest"
5. compile code
6. add the new character (or a blueprint inheriting from it) as default pawn class in the current gamemode and save gamemode file
7. connect 1+ Joycons
8. click play (or start the debugger in visual studio and click play there)

```
UE4Editor_JoyConSample!AMyCharacter::Tick() [C:\Users\...\Documents\Git\UE4-JoyConDriver\Source\JoyConSample\MyCharacter.cpp:35]
UE4Editor_Engine!AActor::TickActor() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:1092]
UE4Editor_Engine!FActorTickFunction::ExecuteTick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:172]
UE4Editor_Engine!FTickFunctionTask::DoTask() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:284]
UE4Editor_Engine!TGraphTask<FTickFunctionTask>::ExecuteTask() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Public\Async\TaskGraphInterfaces.h:886]
UE4Editor_Core!FNamedTaskThread::ProcessTasksNamedThread() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:710]
UE4Editor_Core!FNamedTaskThread::ProcessTasksUntilQuit() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:602]
UE4Editor_Core!FTaskGraphImplementation::WaitUntilTasksComplete() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:1532]
UE4Editor_Engine!FTickTaskSequencer::ReleaseTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:564]
UE4Editor_Engine!FTickTaskManager::RunTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:1585]
UE4Editor_Engine!UWorld::RunTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\LevelTick.cpp:785]
UE4Editor_Engine!UWorld::Tick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\LevelTick.cpp:1494]
UE4Editor_UnrealEd!UEditorEngine::Tick() [D:\Build\++UE4\Sync\Engine\Source\Editor\UnrealEd\Private\EditorEngine.cpp:1736]
UE4Editor_UnrealEd!UUnrealEdEngine::Tick() [D:\Build\++UE4\Sync\Engine\Source\Editor\UnrealEd\Private\UnrealEdEngine.cpp:426]
UE4Editor!FEngineLoop::Tick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\LaunchEngineLoop.cpp:4921]
UE4Editor!GuardedMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Launch.cpp:178]
UE4Editor!GuardedMainWrapper() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:137]
UE4Editor!LaunchWindowsStartup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:273]
UE4Editor!WinMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:320]
UE4Editor!__scrt_common_main_seh() [d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288]
kernel32
ntdll
```

When using the blueprint api instead of c++ one it even works in Unreal Engine 4.27.2 (only needed to change FloatAxis to Axis1D in plugin code), but it crashed when starting the project using switchboard. I had hoped the crashes were due to the same read access violation, but I will do some research on wether or not ndisplay allows for custom input drivers.
If someone has an idea on how to fix the ndisplay crashes with blueprint or c++ api, it would be greatly appreciated.

**Comment by jpdante** [comment]
Hello again, I made some tweaks to add compatibility with Unreal Engine 4.27.x and tried to make the engine not crash when trying to initialize the input plugin or when user code tries to get the plugin or its functions(#4).

Could you please update your plugin and your code and test with the new version?
Please check the wiki or source code to see the new way to how to properly call the API.

Thanks for your interest in my project and for the bug report!

**Comment by jKnepel** [comment]
Hi, the update to the compability worked perfectly. Both blueprint and c++ api works in 4.27 now. The Drivers are even usable with ndisplay, no crashes, nothing. 
However when using the joycons with ndisplay (starting the project over switchboard) there's a ~2sec input delay. The delay is only present for the joycons and not the keyboard/mouse and is the exact same delay time for any button/thumbstick on the joycons. The delay is not present when playing the project in the editor using the same joycons.

If you are able to do something about the delay timer, it would be much appreciated. I am happy to be of assistance with any testing etc. In any case thanks a lot for the great work (and drivers) already!

**Comment by jpdante** [comment]
Hello, I'm going to look at the ~2s delay, only problem is that I don't have any idea of how to use the NDisplay from Unreal, can you try to do some tests and see if you find out anything about that?

**Comment by jKnepel** [comment]
Will do some tests/debugging. If you need any configuration files etc. for setting up ndisplay just message me.

Ndisplay projects are are pretty easily created by choosing filmmaking > ndisplay in the templates or by activating the ndisplay plugin and creating a ndisplay config file. A single node should be sufficient for testing purposes.

After creating an ndisplay file simply start switchboard and switchboard listener and connect the nodes. (127.0.0.1 for node ips work perfectly for testing)

**Comment by jKnepel** [comment]
I've tinkered with the ndisplay setup and the _Low Latency Frame Syncing_ for a bit and I managed to get the delay to pretty much 0. The delay differs a lot depending on the Setup/Computer, but that might be bluetooth related. The tiny delay that always remains will likely be due to either bluetooth connection or joycon drift, so I think we can consider this issue closed.
Bluetooth and C++ API works perfectly now, even in 4.27.x and using the drivers with ndisplay works without any complaints.

Thanks a lot for your amazing drivers and quick help, message me if anything else test related or otherwise is needed and have a great weekend!
